pkg/controller/association/controller/apm_es.go

```diff
@@ -9,39 +9,64 @@ import (
 
 	apmv1 "github.com/elastic/cloud-on-k8s/pkg/apis/apm/v1"
 	commonv1 "github.com/elastic/cloud-on-k8s/pkg/apis/common/v1"
+	esv1 "github.com/elastic/cloud-on-k8s/pkg/apis/elasticsearch/v1"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/association"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/common/operator"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/common/version"
+	eslabel "github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/label"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/services"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/user"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
 	"github.com/elastic/cloud-on-k8s/pkg/utils/rbac"
 	pkgerrors "github.com/pkg/errors"
 	"k8s.io/apimachinery/pkg/types"
 	"sigs.k8s.io/controller-runtime/pkg/manager"
 )
 
 const (
-	// ApmESAssociationLabelName marks resources created by this controller for easier retrieval.
-	ApmESAssociationLabelName = "apmassociation.k8s.elastic.co/name"
-	// ApmESAssociationLabelNamespace marks resources created by this controller for easier retrieval.
-	ApmESAssociationLabelNamespace = "apmassociation.k8s.elastic.co/namespace"
+	// ApmAssociationLabelName marks resources created by this controller for easier retrieval.
+	ApmAssociationLabelName = "apmassociation.k8s.elastic.co/name"
+	// ApmAssociationLabelNamespace marks resources created by this controller for easier retrieval.
+	ApmAssociationLabelNamespace = "apmassociation.k8s.elastic.co/namespace"
+	// ApmAssociationLabelNamespace marks resources created by this controller for easier retrieval.
+	ApmAssociationTypeLabelNamespace = "apmassociation.k8s.elastic.co/type"
 )
 
 func AddApmES(mgr manager.Manager, accessReviewer rbac.AccessReviewer, params operator.Parameters) error {
 	return association.AddAssociationController(mgr, accessReviewer, params, association.AssociationInfo{
-		AssociatedShortName:   "apm",
-		AssociatedObjTemplate: func() commonv1.Associated { return &apmv1.ApmServer{} },
-		AssociationName:       "apm-es",
+		AssociatedShortName:    "apm",
+		AssociationObjTemplate: func() commonv1.Association { return &apmv1.ApmEsAssociation{} },
+		ElasticsearchRef: func(c k8s.Client, association commonv1.Association) (bool, commonv1.ObjectSelector, error) {
+			return true, association.AssociationRef(), nil
+		},
+		ExternalServiceURL: getElasticsearchExternalURL,
+		AssociatedNamer:    esv1.ESNamer,
+		AssociationName:    "apm-es",
 		AssociationLabels: func(associated types.NamespacedName) map[string]string {
 			return map[string]string{
-				ApmESAssociationLabelName:      associated.Name,
-				ApmESAssociationLabelNamespace: associated.Namespace,
+				ApmAssociationLabelName:          associated.Name,
+				ApmAssociationLabelNamespace:     associated.Namespace,
+				ApmAssociationTypeLabelNamespace: "elasticsearch",
```

Comment:
We should probably make `"elasticsearch"` a constant somewhere in `apis/elasticsearch/v1`